prusti-contracts/prusti-std/src/collections.rs

```diff
@@ -0,0 +1,21 @@
+use prusti_contracts::*;
+
+#[extern_spec]
+impl<K, V, S> std::collections::hash_map::HashMap<K, V, S>
+where
+    K: Eq + ::core::hash::Hash,
+    S: ::std::hash::BuildHasher,
+{
+    #[pure]
```

Comment:
Is marking this generic method as `#[pure]` also sound in the presence of interior mutability? (e.g. what if a hypothetical `PartialEq::eq` implementation modifies its internals?)

Reply:
That's a good point, I don't think it would be. It's also related to https://github.com/viperproject/prusti-dev/issues/1286#issuecomment-1406247838.

I wonder if we can restrict this `#[extern_spec]` to types `K` for which `Eq` has been defined structurally, i.e. only using `#[derive(PartialEq, Eq)]` and not via `impl PartialEq/Eq`.

Reply:
Would that also work if `K` is replaced by `&K` which then delegates its impl to a potentially non-structural impl?

Reply:
(It would be nice to add to this PR a test where a key type implements a random `PartialEq` and `Hash`. A client of the `HashMap` should then *not* be able to assert anything about the result of `HashMap::get`.)

Reply:
> Would that also work if `K` is replaced by `&K` which then delegates its impl to a potentially non-structural impl?

I edited my message. The issue is not just interior mutability, but also non-determinism.

Reply:
> I wonder if with `refine_spec` we can restrict all these specifications to types `K/Q` for which `Eq/Hash/Borrow` is pure.

We could.

I think `get` being deterministic is not controversial for a `HashMap`.

However, I think/hope interior mutability isn't actually such a big problem. In my mental model, an interior-mutable structure can be expected to be snapshot equal. As we encode its snapshot recursively, we will eventually find the `UnsafeCell`, which is where we stop: the cell is a *handle*, which does not tell us anything about the value until we also have e.g. a heap (version) to refer to.